File: 01-Two-Pointers/code/javascript/triplet_sum.js
/**
 * Triplet Sum - Two Pointers Solution
 * Author: Ritesh Rana
 * 
 * Problem: Given an array of integers, return all triplets [a, b, c] such that a + b + c = 0.
 * The solution must not contain duplicate triplets.
 * 
 * Time Complexity: O(n²)
 * Space Complexity: O(n) - not counting output space
 */

/**
 * Find all triplets that sum to zero using two-pointer technique.
 * @param {number[]} nums - Array of integers
 * @returns {number[][]} Array of triplets [a, b, c] where a + b + c = 0
 */
function tripletSum(nums) {
    const triplets = [];
    nums.sort((a, b) => a - b);
    
    for (let i = 0; i < nums.length; i++) {
        // Optimization: triplets consisting of only positive numbers will never sum to 0
        if (nums[i] > 0) break;
        
        // To avoid duplicate triplets, skip 'a' if it's the same as the previous number
        if (i > 0 && nums[i] === nums[i - 1]) continue;
        
        // Find all pairs that sum to a target of '-a' ('-nums[i]')
        const pairs = pairSumSortedAllPairs(nums, i + 1, -nums[i]);
        for (const pair of pairs) {
            triplets.push([nums[i], ...pair]);
        }
    }
    
    return triplets;
}

/**
 * Find all pairs in sorted array that sum to target value.
 * @param {number[]} nums - Sorted array of integers
 * @param {number} start - Starting index for the search
 * @param {number} target - Target sum value
 * @returns {number[][]} Array of pairs [b, c] where b + c = target
 */
function pairSumSortedAllPairs(nums, start, target) {
    const pairs = [];
    let left = start, right = nums.length - 1;
    
    while (left < right) {
        const sum = nums[left] + nums[right];
        
        if (sum === target) {
            pairs.push([nums[left], nums[right]]);
            left++;
            
            // To avoid duplicate '[b, c]' pairs, skip 'b' if it's the same as the previous number
            while (left < right && nums[left] === nums[left - 1]) {
                left++;
            }
        } else if (sum < target) {
            left++;
        } else {
            right--;
        }
    }
    
    return pairs;
}

/**
 * Brute force solution for comparison - O(n³) time complexity.
 * @param {number[]} nums - Array of integers
 * @returns {number[][]} Array of triplets that sum to zero
 */
function tripletSumBruteForce(nums) {
    const n = nums.length;
    const triplets = new Set();
    
    for (let i = 0; i < n; i++) {
        for (let j = i + 1; j < n; j++) {
            for (let k = j + 1; k < n; k++) {
                if (nums[i] + nums[j] + nums[k] === 0) {
                    const triplet = [nums[i], nums[j], nums[k]].sort((a, b) => a - b);
                    triplets.add(JSON.stringify(triplet));
                }
            }
        }
    }
    
    return Array.from(triplets).map(triplet => JSON.parse(triplet));
}

/**
 * Test function with comprehensive test cases.
 */
function testTripletSum() {
    console.log("Testing Triplet Sum Solutions");
    console.log("=".repeat(40));
    
    const testCases = [
        {
            input: [0, -1, 2, -3, 1],
            expected: [[-3, 1, 2], [-1, 0, 1]],
            description: 'Basic example'
        },
        {
            input: [],
            expected: [],
            description: 'Empty array'
        },
        {
            input: [0],
            expected: [],
            description: 'Single element'
        },
        {
            input: [1, -1],
            expected: [],
            description: 'Two elements'
        },
        {
            input: [0, 0, 0],
            expected: [[0, 0, 0]],
            description: 'All same values'
        },
        {
            input: [1, 0, 1],
            expected: [],
            description: 'No valid triplets'
        },
        {
            input: [0, 0, 1, -1, 1, -1],
            expected: [[-1, 0, 1]],
            description: 'Duplicate triplets'
        },
        {
            input: [-1, 0, 1, 2, -1, -4],
            expected: [[-1, -1, 2], [-1, 0, 1]],
            description: 'Complex case'
        }
    ];
    
    testCases.forEach((test, index) => {
        // Test optimized solution
        const result = tripletSum([...test.input]);
        const resultSorted = result.map(triplet => [...triplet].sort((a, b) => a - b))
                                   .sort((a, b) => JSON.stringify(a).localeCompare(JSON.stringify(b)));
        const expectedSorted = test.expected.map(triplet => [...triplet].sort((a, b) => a - b))
                                          .sort((a, b) => JSON.stringify(a).localeCompare(JSON.stringify(b)));
        
        const success = JSON.stringify(resultSorted) === JSON.stringify(expectedSorted);
        const status = success ? "✓ PASS" : "✗ FAIL";
        
        console.log(`Test ${index + 1}: ${test.description}`);
        console.log(`Input: [${test.input.join(', ')}]`);
        console.log(`Expected: ${JSON.stringify(test.expected)}`);
        console.log(`Got: ${JSON.stringify(result)}`);
        console.log(`Result: ${status}`);
        
        // Test brute force for comparison (only for small inputs)
        if (test.input.length <= 6) {
            const bruteResult = tripletSumBruteForce([...test.input]);
            const bruteSorted = bruteResult.map(triplet => [...triplet].sort((a, b) => a - b))
                                          .sort((a, b) => JSON.stringify(a).localeCompare(JSON.stringify(b)));
            const bruteMatch = JSON.stringify(bruteSorted) === JSON.stringify(expectedSorted);
            console.log(`Brute force match: ${bruteMatch ? '✓' : '✗'}`);
        }
        
        console.log("-".repeat(40));
    });
}

// Run tests if this file is executed directly
if (typeof require !== 'undefined' && require.main === module) {
    // Run comprehensive tests
    testTripletSum();
    
    // Interactive example
    console.log("\nInteractive Example:");
    const exampleInput = [0, -1, 2, -3, 1];
    console.log(`Input: [${exampleInput.join(', ')}]`);
    const result = tripletSum(exampleInput);
    console.log(`Triplets that sum to zero: ${JSON.stringify(result)}`);
}

// Export functions for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        tripletSum,
        pairSumSortedAllPairs,
        tripletSumBruteForce,
        testTripletSum
    };
}

File: 01-Two-Pointers/code/python/triplet_sum.py
"""
Triplet Sum - Two Pointers Solution
Author: Ritesh Rana

Problem: Given an array of integers, return all triplets [a, b, c] such that a + b + c = 0.
The solution must not contain duplicate triplets.

Time Complexity: O(n²)
Space Complexity: O(n) - not counting output space
"""

from typing import List


def triplet_sum(nums: List[int]) -> List[List[int]]:
    """
    Find all triplets that sum to zero using two-pointer technique.
    
    Args:
        nums: List of integers
        
    Returns:
        List of triplets [a, b, c] where a + b + c = 0
    """
    triplets = []
    nums.sort()
    
    for i in range(len(nums)):
        # Optimization: triplets consisting of only positive numbers will never sum to 0
        if nums[i] > 0:
            break
            
        # To avoid duplicate triplets, skip 'a' if it's the same as the previous number
        if i > 0 and nums[i] == nums[i - 1]:
            continue
            
        # Find all pairs that sum to a target of '-a' ('-nums[i]')
        pairs = pair_sum_sorted_all_pairs(nums, i + 1, -nums[i])
        for pair in pairs:
            triplets.append([nums[i]] + pair)
    
    return triplets


def pair_sum_sorted_all_pairs(nums: List[int], start: int, target: int) -> List[List[int]]:
    """
    Find all pairs in sorted array that sum to target value.
    
    Args:
        nums: Sorted list of integers
        start: Starting index for the search
        target: Target sum value
        
    Returns:
        List of pairs [b, c] where b + c = target
    """
    pairs = []
    left, right = start, len(nums) - 1
    
    while left < right:
        current_sum = nums[left] + nums[right]
        
        if current_sum == target:
            pairs.append([nums[left], nums[right]])
            left += 1
            
            # To avoid duplicate '[b, c]' pairs, skip 'b' if it's the same as the previous number
            while left < right and nums[left] == nums[left - 1]:
                left += 1
                
        elif current_sum < target:
            left += 1 
        else:
            right -= 1
            
    return pairs


def triplet_sum_brute_force(nums: List[int]) -> List[List[int]]:
    """
    Brute force solution for comparison - O(n³) time complexity.
    """
    n = len(nums)
    triplets = set()
    
    for i in range(n):
        for j in range(i + 1, n):
            for k in range(j + 1, n):
                if nums[i] + nums[j] + nums[k] == 0:
                    triplet = tuple(sorted([nums[i], nums[j], nums[k]]))
                    triplets.add(triplet)
    
    return [list(triplet) for triplet in triplets]


def test_triplet_sum():
    """Test function with comprehensive test cases."""
    print("Testing Triplet Sum Solutions")
    print("=" * 40)
    
    test_cases = [
        {
            'input': [0, -1, 2, -3, 1],
            'expected': [[-3, 1, 2], [-1, 0, 1]],
            'description': 'Basic example'
        },
        {
            'input': [],
            'expected': [],
            'description': 'Empty array'
        },
        {
            'input': [0],
            'expected': [],
            'description': 'Single element'
        },
        {
            'input': [1, -1],
            'expected': [],
            'description': 'Two elements'
        },
        {
            'input': [0, 0, 0],
            'expected': [[0, 0, 0]],
            'description': 'All same values'
        },
        {
            'input': [1, 0, 1],
            'expected': [],
            'description': 'No valid triplets'
        },
        {
            'input': [0, 0, 1, -1, 1, -1],
            'expected': [[-1, 0, 1]],
            'description': 'Duplicate triplets'
        },
        {
            'input': [-1, 0, 1, 2, -1, -4],
            'expected': [[-1, -1, 2], [-1, 0, 1]],
            'description': 'Complex case'
        }
    ]
    
    for i, test in enumerate(test_cases, 1):
        # Test optimized solution
        result = triplet_sum(test['input'].copy())
        result_sorted = sorted([sorted(triplet) for triplet in result])
        expected_sorted = sorted([sorted(triplet) for triplet in test['expected']])
        
        success = result_sorted == expected_sorted
        status = "✓ PASS" if success else "✗ FAIL"
        
        print(f"Test {i}: {test['description']}")
        print(f"Input: {test['input']}")
        print(f"Expected: {test['expected']}")
        print(f"Got: {result}")
        print(f"Result: {status}")
        
        # Test brute force for comparison (only for small inputs)
        if len(test['input']) <= 6:
            brute_result = triplet_sum_brute_force(test['input'].copy())
            brute_sorted = sorted([sorted(triplet) for triplet in brute_result])
            brute_match = brute_sorted == expected_sorted
            print(f"Brute force match: {'✓' if brute_match else '✗'}")
        
        print("-" * 40)


if __name__ == "__main__":
    # Run comprehensive tests
    test_triplet_sum()
    
    # Interactive example
    print("\nInteractive Example:")
    example_input = [0, -1, 2, -3, 1]
    print(f"Input: {example_input}")
    result = triplet_sum(example_input)
    print(f"Triplets that sum to zero: {result}")